eth/tracers/native/4byte.go

```diff
@@ -0,0 +1,141 @@
+	"math/big"
+	"strconv"
+	"sync/atomic"
+	"time"
+)
+
+func init() {
+	register("4byte", newFourByteTracer)
+}
+
+// fourByteTracer searches for 4byte-identifiers, and collects them for post-processing.
+// It collects the methods identifiers along with the size of the supplied data, so
+// a reversed signature can be matched against the size of the data.
+//
+// Example:
+//   > debug.traceTransaction( "0x214e597e35da083692f5386141e69f47e973b2c56e7a8073b1ea08fd7571e9de", {tracer: "4byte"})
+//   {
+//     0x27dc297e-128: 1,
+//     0x38cc4831-0: 2,
+//     0x524f3889-96: 1,
+//     0xadf59f99-288: 1,
+//     0xc281d19e-0: 1
+//   }
+type fourByteTracer struct {
+	env               *vm.EVM
+	ids               map[string]int   // ids aggregates the 4byte ids found
+	interrupt         uint32           // Atomic flag to signal execution interruption
+	reason            error            // Textual reason for the interruption
+	activePrecompiles []common.Address // Updated on CaptureStart based on given rules
+}
+
+// newFourByteTracer returns a native go tracer which collects
+// 4 byte-identifiers of a tx, and implements vm.EVMLogger.
+func newFourByteTracer() tracers.Tracer {
+	t := &fourByteTracer{
+		ids: make(map[string]int),
+	}
+	return t
+}
+
+// logic borrowed from newJsTracer in eth/tracers/js/tracer.go
+func (t *fourByteTracer) isPrecompiled(addr common.Address) bool {
+	for _, p := range t.activePrecompiles {
+		if p == addr {
+			return true
+		}
+	}
+	return false
+}
+
+// store saves the given identifier and datasize.
+func (t *fourByteTracer) store(id []byte, size int) {
+	key := bytesToHex(id) + "-" + strconv.Itoa(size)
+	t.ids[key] += 1
+}
+
+// CaptureStart implements the EVMLogger interface to initialize the tracing operation.
+func (t *fourByteTracer) CaptureStart(env *vm.EVM, from common.Address, to common.Address, create bool, input []byte, gas uint64, value *big.Int) {
+	t.env = env
+
+	// Update list of precompiles based on current block
+	rules := env.ChainConfig().Rules(env.Context.BlockNumber)
+	t.activePrecompiles = vm.ActivePrecompiles(rules)
+
+	// Save the outer calldata also
+	if len(input) >= 4 {
+		t.store(input[0:4], len(input)-4)
+	}
+}
+
+// CaptureState implements the EVMLogger interface to trace a single step of VM execution.
+func (t *fourByteTracer) CaptureState(pc uint64, op vm.OpCode, gas, cost uint64, scope *vm.ScopeContext, rData []byte, depth int, err error) {
+}
+
+// CaptureEnter is called when EVM enters a new scope (via call, create or selfdestruct).
+func (t *fourByteTracer) CaptureEnter(typ vm.OpCode, from common.Address, to common.Address, input []byte, gas uint64, value *big.Int) {
+	// Skip if tracing was interrupted
+	if atomic.LoadUint32(&t.interrupt) > 0 {
+		t.env.Cancel()
+		return
+	}
+	// Skip any pre-compile invocations, those are just fancy opcodes
+	if t.isPrecompiled(to) {
+		return
+	}
+	if len(input) >= 4 {
+		t.store(input[0:4], len(input)-4)
+	}
```

Comment:
Might be slightly faster if you do the quick `len(input)` check before the `isPrecompile`, which is slightly more costly. So
```suggestion
	if len(input) < 4 {
		return
	}	
	// Skip any pre-compile invocations, those are just fancy opcodes
	if t.isPrecompiled(to) {
		return
	}
	t.store(input[0:4], len(input)-4)
```